internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,114 @@
+/*
+ *
+ * Copyright 2017 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package delegatingresolver implements the default resolver that creates child
+// resolvers to resolver targetURI as well as proxy address.
+package delegatingresolver
+
+import (
+	"fmt"
+	"os"
+
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+type delegatingResolver struct {
+	target         resolver.Target
+	cc             resolver.ClientConn
+	isProxy        bool
+	targetResolver resolver.Resolver
+	targetAddrs    []resolver.Address
+}
+
+type innerClientConn struct {
+	cc           resolver.ClientConn
+	parent       *delegatingResolver
+	resolverType string
+}
+
+func New(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions, targetResolverBuilder resolver.Builder) (resolver.Resolver, error) {
+	var err error
+	r := &delegatingResolver{
+		target: target,
+		cc:     cc,
+	}
+	proxyURI := os.Getenv("HTTPS_PROXY")
+	if proxyURI == "" {
+		proxyURI = os.Getenv("HTTP_PROXY")
+	}
+	r.isProxy = true
+	if proxyURI == "" {
+		r.isProxy = false
+	}
+
+	r.targetResolver, err = targetURIResolver(target, cc, opts, targetResolverBuilder, r)
+
+	if err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+func targetURIResolver(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions, targetResolverBuilder resolver.Builder, resolver *delegatingResolver) (resolver.Resolver, error) {
```

Comment:
Given that there is a single caller for this function, this could simply be inlined at the callsite.